### PR TITLE
Implement #32347 - (RichTextLabel's Item's should be structs)

### LINKED ISSUE #32347: RichTextLabel's Item and Item* structs should not inherit Object, it's too expensive
**Godot version:**
3.2 master (e426169e4ce8cbb0a8c0fb04f75e60976058177e)

**OS/device including version:**
All

**Issue description:**
#23658 changed `RichTextLabel`'s `struct Item` to an `Object`-derived class (and thus all other `Item*` structs that derive from it, including `ItemFX` and its own inheriting structs).

Then #32202 changed the classes inheriting `Item` to also be `Object` and use `GDCLASS()` to fix casting issues.

@reduz says that it's a bad idea and will have a significant performance impact, as `Object` is quite expensive:
```
16:07 <Akien> Next: #32299
16:07 <IssueBot> #32299: fix area2d monitorable changes | https://git.io/JesFq
16:08 <Akien> reduz: We've had quite a few PRs like this lately where people need to `dynamic_cast` a non-Object class or a struct, and so far we've worked this around by making them Object to use Object::cast_to
16:09 <Akien> Is that the right approach? Another example (merged): #32202
16:09 <IssueBot> #32202: Fix crash on some platforms in RichTextLabel. | https://git.io/Je3Gq
16:10 <reduz> Akien: in those cases there is usually an enum where you can ask what it is
16:10 <reduz> Akien: if not, i suppose it could be added
16:11 <reduz> Akien: oh yeah that is a terrible idea, that will send the performance down the sewer
16:11 <reduz> Akien: ask people to use an enum and check the enum, dont ask them to use Object derived classes
16:11 <reduz> Akien: so yes, that PR will need to be reverted
16:12 <reduz> Object is way too expensive to use for simple things, it should never be used in these situations
16:18 <Akien> reduz: I guess this was also not a good idea then? https://github.com/godotengine/godot/pull/23658/files#diff-8333910cf7dd8405c864e7b46fb3b145R106
16:23 <reduz> yeah i guess not
16:24 <reduz> ItemCustomFX  should simply contain the reference to RichTextEffect
```

And https://github.com/godotengine/godot/pull/32202#issuecomment-535043254

> There is an enum so you can do static cast, so dont use Object, this is overkill as Object / GDCLASS is extremely expensive to be used in a case like this. This PR will have to be reverted.

In the end I decided not to revert the PR as it would reintroduce a bad crashing condition, but we need to refactor this to avoid the use of Object for `Item*`.

Comment by @Relintai: https://github.com/godotengine/godot/issues/32103#issuecomment-535114365

> #23658 will need to be heavily changed, because of the inheritance introduced. Namely `Item->ItemFX->(new classes)`, and the pr itself uses this fact quite a bit. This was also the cause why I changed my pr from using static_casts, to begin with.
> 
> I'll think about how to solve that inheritance issue properly. I don't just want to split up the type enum brainlessly. Even though that's probably what I'd do if I did this for myself :P
> 
> By the way please wait with reverting #32202, at least until there is a proper fix for this because when I was debugging this I was able to get RichTextLabels to crash, just by setting their text at runtime to a short text with a number and a newline in it. It was along the lines of "aaa0a\naa". (With BBCode disabled)
> 
> In fact that was the reason I got crashes in the first place, my game has spells, and I use RichTextLabels to show their description, which is just a few lines of text with a few numbers.

CC @Eoin-ONeill-Yokai @Relintai @reduz

### PR DESCRIPTION
**Pull request by Relintai**
To be honest I'm kind of shocked how fast this came together. (I expected I'll need at least 3-ish really hard days.)

Also I found a few smaller things I fixed those as well.

I'll continue trying to break it for a few days.

Please tell if you like this solution or not, I'll change the implementation accordingly.

**Complete list of changes:**

I kept the inheritance tree unchanged, the ItemFX struct isn't in the enum. I replaced the query that gathers every ItemFX struct to just check for the relevant types. I don't think it worths it to add any new variable to check for this at the moment, however if the number of classes start to grow it might worth to add something.

CharFXTransforms now store a ref to the relevant RichTextEffect (as per @reduz), and since it doesn't need that string identifier anymore, I removed that. (I kept the lookup function, as it's useful when setting up the classes.)

The original code constantly allocated CharFXTransforms, now these are only created once, and stored in ItemCustomFX. This way ItemCustomFX doesn't need the environment dictionary, since it was only there so the code can give it to CharFXTransform on creation.

I found an emty for loop, I removed it.

I made _get_custom_effect_by_code return early, as there shouldn't be more than one bbcode per type.

Implements #32347

*Bugsquad edit:* Closes #32347.

### CODE REVIEW

**Comment by Eoin-ONeill-Yokai** [comment]
This seems to break the example project @ https://github.com/Eoin-ONeill-Yokai/Godot-Rich-Text-Effect-Test-Project

Have you changed the interface for getting rich text effects working / have a working example?

**Comment by Relintai** [comment]
Interesting they work for me, I'll check that test project.

**Comment by Eoin-ONeill-Yokai** [comment]
> Interesting they work for me, I'll check that test project.
> 
> Also I haven't changed the interfaces.
> 
> Edit: For me your test project works properly, I'll try to clean the scons cache and rebuild, just to make sure.

Ah ok let me try it again. It could be my scons cache.

**Comment by Eoin-ONeill-Yokai** [comment]
Yes, so far so good but there's one regression regarding ItemFade not working. Desired results should be as they are shown in the original PR.

Thanks for putting in the time to work on this. I was about to sit down and start working on this when your PR came in. :P

File: scene/gui/rich_text_label.cpp
Comment (Eoin-ONeill-Yokai) [inline scene/gui/rich_text_label.cpp]: The fade issue is likely a regression from the _find_by_type function not returning ITEM_FADE types properly?

File: scene/gui/rich_text_label.cpp
Comment (Eoin-ONeill-Yokai) [inline scene/gui/rich_text_label.cpp]: Be careful here. This branch of the code is used to provide shake with it's random direction per line -- NULL pointers are bypassing this and shake will behave not as intended.

**Review by Eoin-ONeill-Yokai — COMMENTED** [review]
I'm not sure if it's a faux-pas to leave a review as a non-core developer, but here are some quick notes on the issues.

I also found a lingering NULL assignment from your original replacing of `Object::cast_to` I believe, which will affect shake randomness per line iirc.

File: scene/gui/rich_text_label.cpp
Comment (Relintai) [reply scene/gui/rich_text_label.cpp]: it's fine, because of the if at line 903. ItemShake is handled after that.

File: scene/gui/rich_text_label.cpp
Comment (Relintai) [reply scene/gui/rich_text_label.cpp]: Yeah this will be the issue.

File: scene/gui/rich_text_label.cpp
Comment (Relintai) [reply scene/gui/rich_text_label.cpp]: Okay I pretty sure I misunderstood what you said first. So what do you mean by this?

Comment (Eoin-ONeill-Yokai) [reply scene/gui/rich_text_label.cpp]: No actually, having looked at your branch more you're correct. It just wasn't super clear when I was previewing the diff.

File: scene/gui/rich_text_label.cpp
Comment (Relintai) [reply scene/gui/rich_text_label.cpp]: Fixed now.

**Comment by akien-mga** [comment]
> I'm not sure if it's a faux-pas to leave a review as a non-core developer, but here are some quick notes on the issues.

It's completely fine to leave reviews, it helps me a lot to decide whether to merge a PR or if further work is needed.

**Comment by akien-mga** [comment]
> (for github: Implements #32347)

Note: You can use `"Fixes #32347` or `"Closes #32347` so that GitHub auto-closes the mentioned issue when merged, see https://help.github.com/en/articles/closing-issues-using-keywords

**Review by akien-mga — APPROVED** [review]
Looks good to me, can be merged once the updated PR is approved by @Eoin-ONeill-Yokai.

**Comment by Relintai** [comment]
> Note: You can use `"Fixes #32347` or `"Closes #32347` so that GitHub auto-closes the mentioned issue when merged, see https://help.github.com/en/articles/closing-issues-using-keywords

My bad, I've read this a few day ago, and for some reason I thought implements works as well.

**Review by Eoin-ONeill-Yokai — APPROVED** [review]
This looks good, sorry for the delay.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
